File: 1590.py
class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        partials = []
        tot = 0
        n = len(nums)
        for i in range(n):
            tot += nums[i]
            tot %= p
            partials.append(tot)
        target = partials[-1]
        if target == 0:
            return 0
        best = n
        recents = {0:-1}
        for j in range(n):
            if nums[j] == target:
                return 1
            diff = (partials[j] - target) % p
            if diff in recents:
                best = min(best, j - recents[diff])
            recents[partials[j] % p] = j
        if best == n:
            return -1
        return best

File: code/1590.cpp
class Solution {
public:
    int minSubarray(vector<int>& nums, int p) {
        int sum = 0;
        for (int i = 0; i < nums.size(); i++) {
            sum += nums[i];
            sum %= p;
        }
        if (sum == 0) {
            return 0;
        }
        unordered_map<int, int> last_seen;
        last_seen[0] = -1;
        int acc = 0;
        int best = nums.size();
        for (int j = 0; j < nums.size(); j++) {
            acc += nums[j];
            acc %= p;
            int target = acc - sum + p;
            target %= p;
            if (last_seen.contains(target)) {
                best = min(best, j - last_seen[target]);
            }
            last_seen[acc] = j;
        }
        if (best == nums.size()) {
            return -1;
        }
        return best;
    }
};
